csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3521 @@
+          auto def0 = def_group_0->front();
+          auto def1 = def_group_1->front();
+          if (exprsMap(def0, def1, false)) {
+            mapExprs(def0, def1);
+            mapThroughExpr(def0, def1, false);
+          }
+        }
+      }
+    }
+  }
+}
+
+bool IdGraph::mapThroughExpr(Expr* first, Expr* second, bool forward) {
+  if (first == nullptr || second == nullptr) {
+    return false;
+  }
+
+  if (!exprsMap(first, second, forward)) {
+    return false;
+  }
+
+  auto first_ids = ir_utils::filterByType<IterDomain>(
+                       forward ? first->outputs() : first->inputs())
+                       .vector();
+  auto second_ids = ir_utils::filterByType<IterDomain>(
+                        forward ? second->outputs() : second->inputs())
+                        .vector();
+  TORCH_INTERNAL_ASSERT(
+      first_ids.size() == second_ids.size(),
+      "This should be unreachable, if transformation expressions match, their number of inputs and outputs should as well.\n However found:\n",
+      first->toString(),
+      "\nand\n",
+      second->toString());
+  for (auto out_i : c10::irange(first_ids.size())) {
+    mapIds(first_ids[out_i], second_ids[out_i]);
+  }
+
+  return true;
+}
+
+void IterDomainGraphs::assertNoSelfMapping() {
+  TORCH_INTERNAL_ASSERT(
+      !hasSelfMapping(),
+      "Unsupported domain mapping detected in ",
+      std::get<0>(*self_mapping_info_)->toString(),
+      ". ",
+      std::get<3>(*self_mapping_info_),
+      " domains, ",
+      std::get<1>(*self_mapping_info_)->toString(),
+      " and ",
+      std::get<2>(*self_mapping_info_)->toString(),
+      ", are mapped with each other.");
+}
+
+void IdGraph::mapThroughTrivialExprs() {
+  // Grab all expressions
+  std::vector<Expr*> exprs;
+
+  for (auto expr_group : disjointExprSets().disjointSets()) {
+    for (auto expr : *expr_group) {
+      exprs.push_back(expr);
+    }
+  }
+
+  for (auto expr : exprs) {
+    // If not trivial continue
+    auto mapped_ids = IdGraph::isTrivialExpr(expr);
+    if (mapped_ids.empty()) {
+      continue;
+    }
+
+    // Map through trivial expressions
+    for (auto mapped_id_group : mapped_ids) {
+      for (auto id : mapped_id_group) {
+        mapIds(mapped_id_group.front(), id);
+      }
+    }
+  }
+}
+
+void IdGraph::removeTrivialExprs() {
+  ExprGroups trivial_expr_groups;
+  for (auto expr_group : disjointExprSets().disjointSets()) {
+    auto inp_groups = inputGroups(expr_group);
+    auto out_groups = outputGroups(expr_group);
+    if (inp_groups.intersect(out_groups).size()) {
+      trivial_expr_groups.pushBack(expr_group);
+    }
+  }
+
+  // Clear out expressions that map inputs and outputs to the same group
+  // from definitions and uses. They shouldn't be important in traversal, and
+  // will break the terminal input/terminal output logic of traversal. Similar
+  // to what's drafted in buildIndexMap
+  for (auto trivial_expr_group : trivial_expr_groups) {
```

Comment:
Minor comment: You could combine this with the previous loop to remove the need for `trivial_expr_groups`.